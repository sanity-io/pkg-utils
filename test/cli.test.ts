import path from 'path'
import {describe, expect, test} from 'vitest'
import {_spawnProject} from './env'

test('should build `js` package', async () => {
  const project = await _spawnProject('js')

  await project.install()

  const {stdout} = await project.run('build')

  expect(stdout).toContain('./src/index.js -> ./dist/index.esm.js')
  expect(stdout).toContain('./src/index.js -> ./dist/index.js')

  await project.remove()
})

test('should build `exports-dummy` package', async () => {
  const project = await _spawnProject('exports-dummy')

  await project.install()

  const {stdout} = await project.run('build')

  // types
  expect(stdout).toContain('exports-dummy: ./src/index.ts -> ./dist/src/index.d.ts')
  expect(stdout).toContain('exports-dummy/extra: ./src/extra.ts -> ./dist/src/extra.d.ts')

  // commonjs
  expect(stdout).toContain('exports-dummy: ./src/index.ts -> ./dist/index.cjs')
  expect(stdout).toContain('exports-dummy: ./src/index.ts -> ./dist/index.browser.cjs')
  expect(stdout).toContain('exports-dummy/extra: ./src/extra.ts -> ./dist/extra.cjs')
  expect(stdout).toContain('exports-dummy/extra: ./src/extra.ts -> ./dist/extra.browser.cjs')

  // esm
  expect(stdout).toContain('exports-dummy: ./src/index.ts -> ./dist/index.js')
  expect(stdout).toContain('exports-dummy: ./src/index.ts -> ./dist/index.browser.js')
  expect(stdout).toContain('exports-dummy/extra: ./src/extra.ts -> ./dist/extra.js')
  expect(stdout).toContain('exports-dummy/extra: ./src/extra.ts -> ./dist/extra.browser.js')

  expect(await project.readFile('dist/src/index.d.ts')).toMatchSnapshot()
  expect(await project.readFile('dist/index.cjs')).toMatchSnapshot()
  expect(await project.readFile('dist/index.js')).toMatchSnapshot()
  expect(await project.readFile('dist/index.browser.js')).toMatchSnapshot()

  expect(await project.readFile('dist/src/extra.d.ts')).toMatchSnapshot()
  expect(await project.readFile('dist/extra.cjs')).toMatchSnapshot()
  expect(await project.readFile('dist/extra.js')).toMatchSnapshot()
  expect(await project.readFile('dist/extra.browser.js')).toMatchSnapshot()

  await project.remove()
})

test('should build `custom-dist` package', async () => {
  const project = await _spawnProject('custom-dist')

  await project.install()

  const {stdout} = await project.run('build')

  expect(stdout).toContain('./src/index.ts -> ./lib/index.cjs')
  expect(stdout).toContain('./src/index.ts -> ./lib/index.js')
  expect(stdout).toContain('./src/index.ts -> ./lib/src/index.d.ts')

  expect(await project.readFile('lib/index.cjs')).toMatchSnapshot()
  expect(await project.readFile('lib/index.js')).toMatchSnapshot()
  expect(await project.readFile('lib/src/index.d.ts')).toMatchSnapshot()

  await project.remove()
})

test('should build `multi-export` package', async () => {
  const project = await _spawnProject('multi-export')

  await project.install()

  const {stdout} = await project.run('build')

  expect(stdout).toContain('./src/index.ts -> ./dist/index.cjs')
  expect(stdout).toContain('./src/index.ts -> ./dist/index.js')
  expect(stdout).toContain('./src/index.ts -> ./dist/src/index.d.ts')

  expect(stdout).toContain('./src/plugin.ts -> ./dist/plugin.cjs')
  expect(stdout).toContain('./src/plugin.ts -> ./dist/plugin.js')
  expect(stdout).toContain('./src/plugin.ts -> ./dist/src/plugin.d.ts')

  expect(await project.readFile('dist/index.cjs')).toMatchSnapshot()
  expect(await project.readFile('dist/index.js')).toMatchSnapshot()
  expect(await project.readFile('dist/src/index.d.ts')).toMatchSnapshot()

  expect(await project.readFile('dist/plugin.cjs')).toMatchSnapshot()
  expect(await project.readFile('dist/plugin.js')).toMatchSnapshot()
  expect(await project.readFile('dist/src/plugin.d.ts')).toMatchSnapshot()

  await project.remove()
})

test('should build `ts` package', async () => {
  const project = await _spawnProject('ts')

  await project.install()

  const {stdout} = await project.run('build')

  expect(stdout).toContain('./src/index.ts -> ./dist/src/index.d.ts')

  expect(await project.readFile('dist/src/index.d.ts')).toMatchSnapshot()

  await project.remove()
})

describe.skip('runtime: webpack v3', () => {
  test('import `dist/*.browser.js` from package', async () => {
    const exportsDummy = await _spawnProject('exports-dummy')
    const runtime = await _spawnProject('runtime-webpack-v3')

    // install and build dummy package
    await exportsDummy.install()
    await exportsDummy.run('build')

    await runtime.install()

    // build (uses webpack v3)
    await runtime.run('build')

    // eslint-disable-next-line @typescript-eslint/no-var-requires
    const {index, extra} = runtime.require('dist/main.js')

    expect(index.runtime).toBe('browser')
    expect(index.format).toBe('esm')
    expect(index.path).toBe('dist/index.browser.js')

    expect(extra.runtime).toBe('browser')
    expect(extra.format).toBe('esm')
    expect(extra.path).toBe('dist/extra.browser.js')

    await runtime.remove()
    await exportsDummy.remove()
  })
})

describe.skip('runtime: next.js', () => {
  test('import `dist/*.browser.js` from package', async () => {
    const exportsDummy = await _spawnProject('exports-dummy')
    const runtime = await _spawnProject('runtime-next-js')

    // install and build dummy package
    await exportsDummy.install()
    await exportsDummy.run('build')

    await runtime.install()

    // build (uses `next build`)
    await runtime.run('build')

    const buildManifest = runtime.require('.next/build-manifest.json')

    const pageScriptPath = buildManifest.pages['/'].find((f: string) => f.includes('pages/index'))

    const pageFile = await runtime.readFile(path.join('.next', pageScriptPath))

    expect(pageFile).toContain(`"dist/index.browser.js"`)
    expect(pageFile).toContain(`"dist/extra.browser.js"`)

    await runtime.remove()
    await exportsDummy.remove()
  })
})
