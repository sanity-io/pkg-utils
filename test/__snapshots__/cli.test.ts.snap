// Vitest Snapshot v1

exports[`should build \`custom-dist\` package 1`] = `
"\\"use strict\\";Object.defineProperty(exports,\\"__esModule\\",{value:!0});exports.VERSION=\\"1.0.0\\";
//# sourceMappingURL=index.cjs.map
"
`;

exports[`should build \`custom-dist\` package 2`] = `
"const o=\\"1.0.0\\";export{o as VERSION};
//# sourceMappingURL=index.js.map
"
`;

exports[`should build \`custom-dist\` package 3`] = `
"/** @public */
export declare const VERSION = '1.0.0'

export {}
"
`;

exports[`should build \`exports-dummy\` package 1`] = `
"/** @public */
export declare const format: string

/** @public */
export declare const path: string

/** @public */
export declare const runtime: string

export {}
"
`;

exports[`should build \`exports-dummy\` package 2`] = `
"\\"use strict\\";Object.defineProperty(exports,\\"__esModule\\",{value:!0});exports.format=\\"commonjs\\",exports.path=\\"dist/index.cjs\\",exports.runtime=\\"*\\";
//# sourceMappingURL=index.cjs.map
"
`;

exports[`should build \`exports-dummy\` package 3`] = `
"const s=\\"esm\\",e=\\"*\\",t=\\"dist/index.js\\";export{s as format,t as path,e as runtime};
//# sourceMappingURL=index.js.map
"
`;

exports[`should build \`exports-dummy\` package 4`] = `
"const s=\\"esm\\",e=\\"browser\\",r=\\"dist/index.browser.js\\";export{s as format,r as path,e as runtime};
//# sourceMappingURL=index.browser.js.map
"
`;

exports[`should build \`exports-dummy\` package 5`] = `
"/** @public */
export declare const format: string

/** @public */
export declare const path: string

/** @public */
export declare const runtime: string

export {}
"
`;

exports[`should build \`exports-dummy\` package 6`] = `
"\\"use strict\\";Object.defineProperty(exports,\\"__esModule\\",{value:!0});exports.format=\\"commonjs\\",exports.path=\\"dist/extra.cjs\\",exports.runtime=\\"*\\";
//# sourceMappingURL=extra.cjs.map
"
`;

exports[`should build \`exports-dummy\` package 7`] = `
"const s=\\"esm\\",t=\\"*\\",e=\\"dist/extra.js\\";export{s as format,e as path,t as runtime};
//# sourceMappingURL=extra.js.map
"
`;

exports[`should build \`exports-dummy\` package 8`] = `
"const r=\\"esm\\",s=\\"browser\\",e=\\"dist/extra.browser.js\\";export{r as format,e as path,s as runtime};
//# sourceMappingURL=extra.browser.js.map
"
`;

exports[`should build \`multi-export\` package 1`] = `
"'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
});
function defineConfig(options) {
  return options;
}
function definePlugin(plugin) {
  return plugin;
}
exports.defineConfig = defineConfig;
exports.definePlugin = definePlugin;
//# sourceMappingURL=index.cjs.map
"
`;

exports[`should build \`multi-export\` package 2`] = `
"function defineConfig(options) {
  return options;
}
function definePlugin(plugin) {
  return plugin;
}
export { defineConfig, definePlugin };
//# sourceMappingURL=index.js.map
"
`;

exports[`should build \`multi-export\` package 3`] = `
"/** @public */
export declare interface Config extends Plugin_2 {
  plugins?: Plugin_2[]
}

/** @public */
export declare function defineConfig(options: Config): Config

/** @public */
export declare function definePlugin(plugin: Plugin_2): Plugin_2

/** @public */
declare interface Plugin_2 {
  name: string
}
export {Plugin_2 as Plugin}

export {}
"
`;

exports[`should build \`multi-export\` package 4`] = `
"'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
});
var multiExport = require('multi-export');
function plugin() {
  return multiExport.definePlugin({
    name: \\"plugin\\"
  });
}
exports.plugin = plugin;
//# sourceMappingURL=plugin.cjs.map
"
`;

exports[`should build \`multi-export\` package 5`] = `
"import { definePlugin } from 'multi-export';
function plugin() {
  return definePlugin({
    name: \\"plugin\\"
  });
}
export { plugin };
//# sourceMappingURL=plugin.js.map
"
`;

exports[`should build \`multi-export\` package 6`] = `
"/** @public */
export declare function plugin(): Plugin_2

/** @public */
declare interface Plugin_2 {
  name: string
}

export {}
"
`;

exports[`should build \`ts\` package 1`] = `
"/**
 * @internal
 */
export declare interface _Dummy {
  field: string
}

/**
 * @public
 */
export declare interface IncludedModuleDummy {
  field: string
}

/** @public */
export declare const VERSION = '1.0.0'

export {}

declare module './module2' {
  interface IncludedModuleDummy {
    addedField: string
  }
}
//# sourceMappingURL=module1.d.ts.map
"
`;
